### ISSUE TITLE
Consider if metadata resources can be consolidated to the SIOPv2 Request

### ISSUE THREAD

**Issue by KendallWeihe** [TBD54566975/known-customer-credential#20]
PR https://github.com/TBD54566975/known-customer-credential/pull/12 proposes to add the requirement of two metadata resources hosted at well known URIs

Link specifically to the relevant part of the README: https://github.com/TBD54566975/known-customer-credential/tree/kendallw/credential-iss-metadata?tab=readme-ov-file#1-metadata-endpoints

1. Credential Issuer Metadata ([OID4VCI](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-credential-issuer-metadata)) (`/.well-known/openid-credential-issuer`)
2. OAuth 2.0 Authorization Server Metadata ([OID4VCI](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-11.3)) ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414)) (`/.well-known/oauth-authorization-server`)

Rather than offering the metadata at these two well known URIs, could we instead include either/both inside [the SIOPv2 Authorization Request's `client_metadata`](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-9-5.1)?

Motivation: two less endpoints required to be hosted, and two less requests the client is required to make, would reduce the perception of complexity in our developer guides.

---

## Starting from SIOPv2

From [SIOPv2 Authorization Request](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-9-5.1)

> client_metadata: OPTIONAL. This parameter is used by the RP to provide information about itself to a Self-Issued OP that would normally be provided to an OP during Dynamic RP Registration, as specified in [Section 7.3](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#rp-registration-parameter).

[Section 7.3](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-7.3), references OID4VP, but doesn't specify the exact section (spec is a "draft").

<img width="423" alt="Screenshot 2024-03-14 at 4 20 45 PM" src="https://github.com/TBD54566975/known-customer-credential/assets/3602993/58e661c1-ae5b-4dd1-8dc5-8b1ac40110af">

Presumably, the referred-to section in OID4VP is to [9. ](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-9)[Verifier Metadata (Client Metadata)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-verifier-metadata-client-me)

> To convey Verifier metadata, Client metadata defined in Section 2 of [[RFC7591](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#RFC7591)] is used

[Section 2 of RFC7591 is linked here](https://datatracker.ietf.org/doc/html/rfc7591#section-2), and doesn't seem to be compatible with the two metadata resources required (in the ordered list above). Am I missing something?

---

## Alternative Idea

An alternative idea is to include the metadata along side `credential_offer` and `uri` in the [IDV Request](https://github.com/TBD54566975/known-customer-credential/tree/kendallw/credential-iss-metadata?tab=readme-ov-file#idv-request). The [IDV Request](https://github.com/TBD54566975/known-customer-credential/tree/kendallw/credential-iss-metadata?tab=readme-ov-file#idv-request) isn't part of any of the relevant specs, it's a part of _this_ (known-customer-credential) spec. The sequence of events here is:

1. Wallet makes `GET /idv` http request to the issuer
2. Issuer responds with the [SIOPv2 Request](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-openid-provider-a) in the http response body
3. Wallet makes a `POST /idv` http request to the issuer, wherein the http request body is the [SIOPv2 Response](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-self-issued-openid-provider-au)
4. Issuer responds with the [IDV Request](https://github.com/TBD54566975/known-customer-credential/tree/kendallw/credential-iss-metadata?tab=readme-ov-file#idv-request) in the http response body

So the idea being we place the metadata inside the [IDV Request](https://github.com/TBD54566975/known-customer-credential/tree/kendallw/credential-iss-metadata?tab=readme-ov-file#idv-request) wherein the wallet can subsequently make use of the information, rather than needing to turn back around and make two requests to the well known URIs.

With this idea in mind, do we _still need_ to make the well known URIs a required part of this (known-customer-credential) specification, in order to be compatible with OID4VCI? In other words, we support _both_ metadata inside the [IDV Request](https://github.com/TBD54566975/known-customer-credential/tree/kendallw/credential-iss-metadata?tab=readme-ov-file#idv-request) and also at well known URIs. The benefit here being that we could still simplify our developer guides while being spec compatible.

**Comment by decentralgabe** [comment]
> Rather than offering the metadata at these two well known URIs, could we instead include either/both inside [the SIOPv2 Authorization Request's client_metadata](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#section-9-5.1)?

It does not look like the [metadata the spec defines](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html#name-relying-party-metadata-valu) supports this. Perhaps we could leverage the [dynamic client registration metadata](https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata) specifically with the `request_uris` property...but this seems to not be a normal path.

Also noting that the guidance is that we MUST host those endpoints, so I do not think we can get away from hosting these endpoints without normative language making its way into the spec that sets these as optional and allows for providing them dynamically during a SIOPv2 Request..

**Comment by selfissued** [comment]
I agree with Gabe that, to the extent that we're using the OpenID4VC specs, we should use them as intended.  Interoperability will be improved by doing so.

Reacting to the idea of including this metadata in SIOPv2 requests, the problem with that is that the Client/RP is not authoritative for the Credential Issuer/Authorization Server metadata.  Modularity considerations indicate that each party that is participating in the protocol should supply its own metadata.

**Comment by KendallWeihe** [comment]
Thank you to you both @decentralgabe @selfissued 

Any objections or additional commentary @mistermoe @tomdaffurn? Else I'll mark this ticket as closed
